cdc_rsync_server/unzstd_stream.cc

```diff
@@ -67,6 +53,20 @@ absl::Status UnzstdStream::Read(void* out_buffer, size_t out_size,
       return MakeStatus("EOF with %u bytes input data available",
                         input_.size - input_.pos);
     }
+
+    if (input_.pos == input_.size && output.pos < output.size && !*eof) {
```

Comment:
Why was it necessary to move that block here?

Reply:
It's an optimization. ZSTD_decompressStream may produce new output data even if there's no new input data. Before, UnzstdStream::Read would first read new input data (which might block), and then call ZSTD_decompressStream. Imagine you have 10 compressed bytes that produce 40 uncompressed bytes. Now you read 20 bytes twice. The old code would

- Read 10 compressed bytes from the socket
- ZSTD_decompressStream would then uncompress the 10 bytes into 40 bytes, but only output 20 of them (the other 20 are stuck in internal buffers)
- The next call to UnzstdStream::Read would then again read compressed bytes from the socket. This operation might block for a bit, e.g. when the client is slow to produce output (see AutoFlushAfter).

The new code now
- Tries ZSTD_decompressStream first, but doesn't have any input yet, so does not fill output array
- Reads 10 compressed bytes from the socket
- ZSTD_decompressStream uncompresses the 10 bytes into 40 bytes and outputs 20 of them
- The next call to UnzstdStream::Read now calls ZSTD_decompressStream first and outputs the remaining 20 uncompressed bytes. Nothing is read from the socket.

Reply:
Thanks for the explanation!